services/calculateBlockTvl.js

```diff
@@ -0,0 +1,149 @@
+ */
+const calculateBlockTvl = async (lastBlock) => {
+    await karuraApi.isReady;
+    let header = await karuraApi.derive.chain.bestNumberFinalized();
+
+    let previousHeader;
+    if (lastBlock === undefined) {
+        previousHeader = header - 1;
+        lastBlock = {};
+        lastBlock.header = header - 1;
+    } else {
+        previousHeader = lastBlock.header
+    }
+
+    log(chalk.blue.bold('Start calculateBlockTvl'));
+    log(chalk.blue.bold('DB: ') + `Loaded: Last Block ${previousHeader}`);
+    log(chalk.blue.bold('HEADER: current header number: ') + header);
+
+    // Run logic for each block between the last run block and the current block
+    if (header > previousHeader) {
+        if (header == previousHeader + 1) {
+            log(chalk.blue.bold('IMPORT BLOCK: ') + `#${header}`);
+            extractBlockNumber(header, lastBlock);
+        } else {
+            for (let blockNumber = (previousHeader + 1); blockNumber <= header; blockNumber++) {
+                log(chalk.blue.bold('IMPORT BLOCK: ') + `#${blockNumber}`);
+                await extractBlockNumber(blockNumber);
+            }
+        }
+    } else {
+        log(chalk.blue.bold('ALREADY RUN EXTRACTION FOR THIS BLOCK, returning ') + `#${header}`);
+    }
+};
+
+async function extractBlockNumber(blockNumber) {
+    const blockHash = await karuraApi.rpc.chain.getBlockHash(blockNumber);
+    const signedBlock = await karuraApi.rpc.chain.getBlock(blockHash);
+    await extractBlock(signedBlock.block, blockNumber);
+}
+
+async function extractBlock(block, blockNumber) {
+    log(chalk.blue.bold('API: ') + `Extracting Block: ${block.header.hash}`);
+
+    const timestamp = await karuraApi.query.timestamp.now.at(block.header.hash);
+    log(chalk.blue.bold('API: ') + `Timestamp: ${timestamp.toString()}`);
+
+    await extractBlockDexLiquidities(block, timestamp.toString(), blockNumber);
+    // await extractBlockEvents(block, timestamp);
+}
+
+async function extractBlockDexLiquidities(block, timestamp, blockNumber) {
+    const tradingPairs = config.TOKEN_PAIRS.map((pair) => {
+        const tokenA = karuraApi.registry.createType('CurrencyId', { Token: pair[0] });
+        const tokenB = karuraApi.registry.createType('CurrencyId', { Token: pair[1] });
+        const tradingPair = new TokenPair(Token.fromCurrencyId(tokenA), Token.fromCurrencyId(tokenB)).toTradingPair(karuraApi) 
+        return new Promise((resolve, reject) => {
+            karuraApi.query.dex.liquidityPool.at(block.header.hash, tradingPair)
+            .then(res => {
+                resolve({
+                    pair: pair,
+                    data: res
+                })
+            })
+            .catch(e => {
+                reject(e)
+            });
+        });
+    });
+    await extractBlockValue(tradingPairs, timestamp, blockNumber);
+}
+
+async function extractBlockValue(tradingPairs, timestamp, blockNumber) {
+    const dexBalances = await Promise.all(tradingPairs);
```

Comment:
this is a bit confusing, should do the await on caller side

Reply:
agreed